### Import required dependencies

In [ ]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [69]:
# Create the path and import the data
full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [70]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [71]:
# Check for null values
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [72]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [73]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [74]:
# Drop duplicated rows and verify removal
student_df = student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [75]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [76]:
# Examine the grade column to understand why it is not an int
student_df['grade']

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [77]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace('th', '')

In [78]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(int)

student_df['grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int32

In [79]:
student_df['grade'] = student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [80]:
student_df.head()

student_id      student_name  grade             school_name  reading_score  \
0   103880842     Travis Martin      9    Sullivan High School           59.0   
1    45069750     Michael Brown      9       Dixon High School           94.7   
2    45024902   Gabriela Lucero      9      Wagner High School           89.0   
3    62582498  Susan Richardson      9       Silva High School           69.7   
5    74579444   Cynthia Johnson      9  Montgomery High School           63.5   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
5        76.9     Charter         893368

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [81]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [82]:
# Display the mean math score using the mean function
mean_math_score = student_df['math_score'].mean()
mean_math_score

64.67573326141189

In [83]:
# Store the minimum reading score as min_reading_score
min_reading_score = student_df['math_score'].min()
min_reading_score

3.7

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [85]:
# Use loc to display the grade column
student_df.loc[:, ["grade"]]

grade
0          9
1          9
2          9
3          9
5          9
...      ...
19508     10
19509     12
19511     11
19512     11
19513     12

[14831 rows x 1 columns]

In [88]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[0:3, [3, 4, 5]]


school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [89]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
student_grades_df = student_df.loc[student_df["grade"] == 9]
student_grades_df.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [90]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_score = student_df["reading_score"].min()
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [138]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
only_dixon_and_reading_score = student_df.loc[(student_df["school_name"] == "Dixon High School")]
only_dixon


student_id     student_name  grade        school_name  reading_score  \
1        45069750    Michael Brown      9  Dixon High School           94.7   
32       10848508     Brenda Carey     12  Dixon High School           90.3   
45      103118608     Ryan Charles     10  Dixon High School           71.1   
58       73841348       Karen Hart      9  Dixon High School           83.5   
60       86894214   Jeremiah Smith     10  Dixon High School           59.5   
...           ...              ...    ...                ...            ...   
19401    44052072  Anthony Jackson      9  Dixon High School           62.1   
19435   106226791     Jason Conley      9  Dixon High School           61.0   
19445    54346011      David Henry     10  Dixon High School           43.9   
19466    20633952       Joan Boone     12  Dixon High School           95.2   
19477    14165038    William Smith      9  Dixon High School           69.8   

       math_score school_type  school_budget  
1            73.5     Charter         870334  
32           57.9     Charter         870334  
45           93.1     Charter         870334  
58           89.1     Charter         870334  
60           96.4     Charter         870334  
...           ...         ...            ...  
19401        69.9     Charter         870334  
19435        71.2     Charter         870334  
19445        90.4     Charter         870334  
19466        31.4     Charter         870334  
19477        65.0     Charter         870334  

[1583 rows x 8 columns]

In [214]:
# Find the mean reading score for all students in grades 11 and 12 combined.
student_df.loc[student_df["grade"] == 11, "reading_score"].mean()
            

77.47841726618722

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [171]:
# Use groupby and mean to find the average reading and math scores for each school type.
avg_student_scores_by_school_type = student_df.groupby(["school_type", "reading_score", "math_score"]).mean()
avg_student_scores_by_school_type.mean()

student_id       6.975082e+07
grade            1.035372e+01
school_budget    8.937855e+05
dtype: float64

In [210]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.

total_students_by_school_name = student_df.groupby(["school_name", "student_id"]).count()
total_students_by_school_name.count()

student_df.sort_values(["school_name", "student_id"], ascending=False)


student_id      student_name  grade         school_name  reading_score  \
12446   129959948    Peter Williams      9   Woods High School           34.5   
6665    129924546    Valerie Carter     12   Woods High School           82.0   
16878   129811714    Robert Fleming      9   Woods High School           67.3   
11000   129791753     Angela Mosley     12   Woods High School           67.3   
6529    129654586      Tracy Clarke     10   Woods High School           53.7   
...           ...               ...    ...                 ...            ...   
1940     10836251     Jessica Smith     11  Bowers High School           86.9   
250      10752173      Jesse Miller     12  Bowers High School           92.7   
2883     10638873     Leonard Hayes     11  Bowers High School           88.2   
18675    10138027    James Williams     11  Bowers High School           83.1   
12409    10045550  Ronald Dominguez     11  Bowers High School           72.6   

       math_score school_type  school_budget  
12446        69.3      Public         912243  
6665         65.2      Public         912243  
16878        59.3      Public         912243  
11000        61.2      Public         912243  
6529         60.6      Public         912243  
...           ...         ...            ...  
1940         14.5      Public         848324  
250          41.4      Public         848324  
2883         54.6      Public         848324  
18675        38.9      Public         848324  
12409        59.2      Public         848324  

[14831 rows x 8 columns]

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

In this analysis, I made a few discoveries with the student data given. I collected the data with the os.path.join. While preparing the data, I checked for null values as well as duplicates. I switched grade to an integer so it could show up as a number. Next, I did a summary statistics for the data frame in order to help me calculate the mean math score and minimum reading score. The mean math score was 64.67573326141189 and the minimum reading score was 3.7. Following this, I used the loc system to manipulate the data between different columns. Such as, showing the grade column, 9th grade statistics, and the minimum reading score. I struggled using loc with conditionals. I was able to show the reading scores of Dixon High School students, as well as the average reading score for 11th grade. I had trouble commbining 11th and 12th. Lastly, I compared the data to find the average reading and math scores for each school. I had trouble figuring out how to get the count of student ID's. 